File: core/src/main/java/com/codeabovelab/tpc/core/processor/RegexPredicate.java
package com.codeabovelab.tpc.core.processor;

import com.codeabovelab.tpc.core.text.Text;
import com.codeabovelab.tpc.core.text.TextCoordinates;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 */
public final class RegexPredicate implements RulePredicate {
    private final Pattern pattern;

    public RegexPredicate(String regexp) {
        this.pattern = Pattern.compile(regexp);
    }

    @Override
    public List<TextCoordinates> test(PredicateContext pc, Text text) {
        Matcher matcher = pattern.matcher(text.getData());
        List<TextCoordinates> list = null;
        while(matcher.find()) {
            int offset = matcher.start();
            int len = matcher.end() - offset;
            TextCoordinates coord = text.getCoordinates(offset, len);
            if(list == null) {
                list = new ArrayList<>();
            }
            list.add(coord);
        }
        if(list == null) {
            // we always must return non null value
            return Collections.emptyList();
        }
        return list;
    }
}

File: core/src/main/java/com/codeabovelab/tpc/core/processor/RegexPredicate.kt
package com.codeabovelab.tpc.core.processor

import com.codeabovelab.tpc.text.Text
import com.fasterxml.jackson.annotation.JsonTypeName

import java.util.ArrayList
import java.util.regex.Pattern

/**
 */
@JsonTypeName("RegexPredicate")
class RegexPredicate(
        /**
         * Need as property for json representation
         */
        val pattern: String
): RulePredicate<PredicateResult<*>> {

    private val compiled: Pattern = Pattern.compile(this.pattern, Pattern.MULTILINE or Pattern.DOTALL)

    override fun test(pc: PredicateContext, text: Text): PredicateResult<*> {
        val matcher = compiled.matcher(text.data)
        var list: MutableList<PredicateResult.Entry>? = null
        while(matcher.find()) {
            val start = matcher.start()
            val end = matcher.end()
            if(start == end) {
                // it happens on some patterns
                continue
            }
            val len = end - start
            val coord = text.getCoordinates(start, len)
            if(list == null) {
                list = ArrayList()
            }
            list.add(PredicateResult.Entry(coord))
        }
        return PredicateResult(list.orEmpty())
    }
}
